source/MaterialXCore/Document.cpp

```diff
@@ -99,17 +108,27 @@ class Document::Cache
                             {
                                 NodeGraphPtr nodeGraph = impl->getDocument()->getNodeGraph(nodeGraphString);
                                 if (nodeGraph)
-                                    implementationMap[interface->getQualifiedName(nodeDefString)].push_back(nodeGraph);
+                                    nonFuncNodeDefMap[interface->getQualifiedName(nodeDefString)].push_back(nodeGraph);
                             }
                             else
                             {
-                                implementationMap[interface->getQualifiedName(nodeDefString)].push_back(interface);
+                                nonFuncNodeDefMap[interface->getQualifiedName(nodeDefString)].push_back(interface);
                             }
                         }
                     }
                 }
             }
 
+            // Functional node definitions have precedence over non-functional ones
+            // So append them to the end of the implementation list assocaited 
+            // with any existing nodedef entry (or create a new one if does not exist).
+            //
+            for (const auto& [nodedefKey, appendImplementations] : nonFuncNodeDefMap)
```

Comment:
Search order is strictly set here for definitions: functional defs first, all others after. Usage priority is decided elsewhere but generally it's a "first found" logic (per target). 

For this initial check-in, we only look for child nodegraphs. Child implementations could be added if / when desired.

Reply:
This was a remaining question, and I think your logic makes sense. It also addresses a previous concern from Doug about how to extend implementation associations. 

- If there is no overlap then order does not matter. 
- If there is then we can give non-embedded nodegraph implementation a higher priority by adding them to the implementation list first (so that they are found first when looking for implementations).

Note that this does no affect being able to get all the implementations back.